security/pkg/nodeagent/sds/sdsservice_test.go

```diff
@@ -492,6 +516,157 @@ func TestStreamSecretsMultiplePush(t *testing.T) {
 		t.Fatalf("failed to send push notificiation to proxy %q", conID)
 	}
 	notifyChan <- "receive secret"
+	// Wait for pushSDS to return.
+	time.Sleep(2 * time.Second)
+	conn.Close()
+	if notify := <-notifyChan; notify != "close stream" {
+		t.Fatalf("get unexpected notification. %s", notify)
+	}
+	totalPushVal, err := util.GetMetricsCounterValue(sdsMetrics.totalPush)
+	if err != nil {
+		t.Errorf("Fail to get value from metric totalPush: %v", err)
+	}
+	totalPushVal -= initialTotalPush
+	if totalPushVal != float64(1) {
+		t.Errorf("unexpected metric totalPush: expected 1 but got %v", totalPushVal)
+	}
+}
+
+func testSDSStreamUpdateFailures(t *testing.T, stream sds.SecretDiscoveryService_StreamSecretsClient, proxyID string,
+	notifyChan chan string) {
+	req := &api.DiscoveryRequest{
+		ResourceNames: []string{testResourceName},
+		Node: &core.Node{
+			Id: proxyID,
+		},
+	}
+
+	// Send first request and
+	if err := stream.Send(req); err != nil {
+		t.Errorf("stream.Send failed: %v", err)
+	}
+	resp, err := stream.Recv()
+	if err != nil {
+		t.Errorf("stream.Recv failed: %v", err)
+	}
+	verifySDSSResponse(t, resp, fakePrivateKey, fakeCertificateChain)
+
+	// Send second request as an ACK and wait for notifyPush
+	req.VersionInfo = resp.VersionInfo
+	req.ResponseNonce = resp.Nonce
+	if err = stream.Send(req); err != nil {
+		t.Errorf("stream.Send failed: %v", err)
+	}
+	// Wait for SDS server to process the second request and hold the request.
+	time.Sleep(5 * time.Second)
```

Comment:
any way to poll for this? Anytime we have a sleep we are going to be either at risk of flakes, or make our tests much slower than they need to be